opm/core/grid/RepairZCORN.hpp

```diff
@@ -0,0 +1,728 @@
+        {
+            const auto cornerCnt0 = this->topBelowBottom_.corners;
+
+            const auto ijk = this->active_.getCellIJK(globCell);
+
+            for (const auto& pt : this->zcorn_idx_.pillarPoints(ijk)) {
+                const auto zb = this->zcorn_[pt.bottom];
+                auto&      zt = this->zcorn_[pt.top];
+
+                if (zt > zb) {  // Top below bottom (ZCORN is depth)
+                    zt = zb;
+
+                    this->topBelowBottom_.corners += 1;
+                }
+            }
+
+            this->topBelowBottom_.cells +=
+                this->topBelowBottom_.corners > cornerCnt0;
+        }
+
+        /// Ensure ZCORN values honour restriction that cell's bottom corner
+        /// is not below cell's lower neighbour's upper corner on same
+        /// pillar.
+        ///
+        /// Modifies \code this->zcorn_ \endcode.
+        ///
+        /// \tparam IndexTuple Representation of Cartesian index tuples.
+        ///    Typically \code ActiveCells::IndexTuple \endcode.
+        ///
+        /// \param[in] ijk Cartesian position of particular cell.
+        template <class IndexTuple>
+        void ensureCellBottomNotBelowLowerTop(const IndexTuple& ijk)
+        {
+            const auto below = this->active_.neighbourBelow(ijk);
+
+            if (below >= this->active_.numGlobalCells()) {
+                return;
+            }
+
+            const auto cornerCnt0 = this->bottomBelowLowerTop_.corners;
+
+            const auto& up    = this->zcorn_idx_.pillarPoints(ijk);
+            const auto  d_ijk = this->active_.getCellIJK(below);
+            const auto& down  = this->zcorn_idx_.pillarPoints(d_ijk);
+
+            for (auto n = up.size(), i = 0*n; i < n; ++i) {
+                const auto zbu = this->zcorn_[up  [i].bottom];
+                auto&      ztd = this->zcorn_[down[i].top];
+
+                if (zbu > ztd) { // Bottom below lower top (ZCORN is depth)
+                    ztd = zbu;
+
+                    this->bottomBelowLowerTop_.corners += 1;
+                }
+            }
+
+            this->bottomBelowLowerTop_.cells +=
+                this->bottomBelowLowerTop_.corners > cornerCnt0;
+        }
+
+        /// Determine whether or not input ZCORN array represents elevations
+        /// (values decreasing for increasing layer index) or depths (values
+        /// increasing for increasing layer index).
+        bool zcornIsElevation() const
+        {
+            auto all_signs = std::vector<int>{};
+            all_signs.reserve(this->active_.numGlobalCells());
+
+            for (const auto& globCell : this->active_.activeGlobal()) {
+                all_signs.push_back(this->getZCornSign(globCell));
+            }
+
+            const int ignore = 0;
+
+            // Elevation == ZCORN decreasing => all non-ignored cell signs
+            // negative.
+            return (first(all_signs, ignore) == -1)
+                && allEqual(all_signs, ignore, -1);
```

Comment:
Cannot this return statement be replaced with just `return allEqual(all_signs, ignore, -1);`?

Also: what should happen if all_signs contains a mix of signs? It seems to me that we will then assume zcorn is depth.